### PR TITLE
Fixes #10169: UI to define node properties

### PR DESCRIPTION
**Pull request by RaphaelGauthier**
https://www.rudder-project.org/redmine/issues/10169

### CODE REVIEW

File: rudder-web/src/main/resources/demo-rudder-users.xml [hunk 1]
```diff
@@ -74,5 +74,6 @@
 
   <!-- Default credentials: admin / admin -->
   <user name="admin" password="c7ad44cbad762a5da0a452f9e854fdc1e0e7a52a38015f23f3eab1d80b931dd472634dfac71cd34ebc35d16ab7fb8a90c81f975113d6c7538dc69dd8de9077ec" role="administrator" />
+  <user name="readonly" password="c7ad44cbad762a5da0a452f9e854fdc1e0e7a52a38015f23f3eab1d80b931dd472634dfac71cd34ebc35d16ab7fb8a90c81f975113d6c7538dc69dd8de9077ec" role="read_only" />
```
Comment (VinceMacBuche) [inline rudder-web/src/main/resources/demo-rudder-users.xml]: be careful to what you commit

File: rudder-web/src/main/scala/com/normation/rudder/web/services/DisplayNode.scala [hunk 2]
```diff
@@ -655,21 +655,34 @@ object DisplayNode extends Loggable {
...
+      val jsonProperties = compactRender(node.properties.toApiJson())
+      val userHasRights = CurrentUser.checkRights(Write("node"))
+      def tabProperties = ChooseTemplate(List("templates-hidden", "components", "ComponentNodeProperties") , "nodeproperties-tab")
+
+      val css: CssSel =  "#tabPropsId [id]" #> htmlId(jsId,"sd_props_")
+      css(tabProperties) ++ Script(OnLoad(JsRaw(s"""
+        angular.bootstrap('#nodeProp', ['nodeProperties']);
+        var scope  = angular.element($$("#nodeProp")).scope();
+        var props  = ${jsonProperties};
+        var nodeId = "${nodeId}"
```
Comment (VinceMacBuche) [inline rudder-web/src/main/scala/com/normation/rudder/web/services/DisplayNode.scala]: dangerous to create those variables here they will be sent into global scope, maybe modifying global variabled used 

either put the values directly (no vars) or encapsulate into a scope/function

File: rudder-web/src/main/resources/demo-rudder-users.xml (hunk 1 above)
Comment (RaphaelGauthier) [inline rudder-web/src/main/resources/demo-rudder-users.xml]: Oops ..

File: rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+*************************************************************************************
+*/
+
+var app = angular.module('nodeProperties', ['DataTables', 'monospaced.elastic']);
+
+app.controller('nodePropertiesCtrl', function ($scope, $http, $compile) {
+  //Initialize scope
+  $scope.properties;
+  $scope.nodeId;
+  $scope.idTable          = "#nodePropertiesTab";
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js]: tableId

**Comment by RaphaelGauthier** [comment]
Commit modified

File: rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    "fnCreatedCell" : function (nTd, sData, oData, iRow, iCol) {
+      var pre = $("<pre class='json-beautify' onclick='$(this).toggleClass(\"toggle\")'></pre>");
+      var content = sData != null && typeof sData === 'object' ? JSON.stringify(sData) : sData;
+      $(nTd).empty();
+      pre.html(content)
+      $(nTd).prepend(pre);
+     }
+   }
+  , {
+    "aTargets":[2],
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js]: Sad about the third column (you can see it in the table header )

Maybe one day we will not use anymore the datatable directive`




```html
 <table><thead>...</thead><tbody><tr ng-repeat="property in properties">...<tr></tbody></table)
```

File: rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    $scope.overrideOptions = overrideOptions;
+    //Get current node properties
+    $scope.properties = properties;
+    $scope.urlAPI = contextPath + '/secure/api/latest/nodes/'+ nodeId;
+    $($scope.idTable).dataTable().fnAddData($scope.properties);
+  }
+
+  $scope.addProperty = function(){
+    function getIndex(property, index, array) {
+      return property.name == $scope.newProperty.name;
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js]: always === not ==

File: rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    $scope.columnDefs = columnDefs;
+    $scope.overrideOptions = overrideOptions;
+    //Get current node properties
+    $scope.properties = properties;
+    $scope.urlAPI = contextPath + '/secure/api/latest/nodes/'+ nodeId;
+    $($scope.tableId).dataTable().fnAddData($scope.properties);
+  }
+
+  $scope.addProperty = function(){
+    function getIndex(property, index, array) {
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js]: Function does not getIndex, but is checkNameUnicity :)

File: rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+var app = angular.module('nodeProperties', ['DataTables', 'monospaced.elastic']);
+
+app.controller('nodePropertiesCtrl', function ($scope, $http, $compile) {
+  //Initialize scope
+  $scope.properties;
+  $scope.nodeId;
+  $scope.tableId          = "#nodePropertiesTab";
+  $scope.urlAPI           = contextPath + '/secure/api/latest/nodes/';
+  $scope.newProperty      = {'name':"", 'value':""}
+  $scope.selectedProperty = {'name':"", 'index':""};
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js]: Maybe Should be "deletedProperty", don't you think ?

File: rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js (hunk 4 above)
Comment (RaphaelGauthier) [inline rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js]: Yes, I looked to hide it but it is not easily feasible with datatable directive ...
I hope !

File: rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js (hunk 6 above)
Comment (RaphaelGauthier) [inline rudder-web/src/main/webapp/javascript/rudder/angular/nodeProperties.js]: True, I forgot to rename it !

**Comment by RaphaelGauthier** [comment]
Commit modified

**Pull request merged**